### PR TITLE
Drop support for clang 6 and 7

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

We don't rely on clang 6 on any supercomputer and clang 7 was released over 2 years ago. Dropping it now because it has a bug that blocks #2656.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

File: docs/Installation/Installation.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ environment differently, read on!
 #### Required:
 
 * [GCC](https://gcc.gnu.org/) 7.0 or later,
-[Clang](https://clang.llvm.org/) 5.0 or later, or AppleClang 6.0 or later
+[Clang](https://clang.llvm.org/) 7.0 or later, or AppleClang 7.0 or later
```
Comment (kidder) [inline docs/Installation/Installation.md]: I think you need at least AppleClang10.0.1 to be comparable to clang 7.0.  Have you built with AppleClang? I will try to do so in a bit...

**Review by kidder — CHANGES_REQUESTED** [review]

File: docs/Installation/Installation.md (hunk 1 above)
Comment (nilsvu) [reply docs/Installation/Installation.md]: I'm building with AppleClang 12 these days, and built with AppleClang 11 before. Haven't tried an older version for a long time.

File: containers/Dockerfile.buildenv [hunk 2]
```diff
@@ -39,7 +39,7 @@ RUN apt-get update -y \
                           libopenblas-dev liblapack-dev \
                           libhdf5-dev hdf5-tools \
                           libgsl0-dev \
-                          clang-6.0 clang-8 clang-9 \
+                          clang-6.0 clang-7 clang-8 clang-9 \
```
Comment (fmahebert) [inline containers/Dockerfile.buildenv]: why not remove clang-6 here?

Comment (nilsvu) [reply containers/Dockerfile.buildenv]: To avoid breaking everyone's PRs :) I'll remove clang-6 in a few weeks

**Review by wthrowe — CHANGES_REQUESTED** [review]
The cmake compiler check needs to be updated too.

Has anyone tested this with clang 7?  I can do it later.

**Comment by nilsdeppe** [comment]
I have not tested with clang 7. It looks like there was a push to fix things clang was unhappy with.

**Comment by wthrowe** [comment]
This compiles with clang-7.0.0 and libstdc++-10.2.0, but the changes from #2656 still do not.

**Comment by nilsdeppe** [comment]
Okay, clang-6 is the default on Ubuntu 18.04, so by requiring clang-7 we already break that. One can easily install clang-8 on Ubuntu 18.04 so we can still support a ~3-year-old OS while requiring clang-8. I don't know what this would mean for macOS, though. @wthrowe does #2656 work with clang-8?

**Comment by wthrowe** [comment]
I haven't tested personally, but CI passed for it.

**Comment by nilsdeppe** [comment]
Seems reasonably indicative that it works :P 

Okay, I would be fine requiring clang-8 and newer, but I think we should get some feedback from @kidder, @fmahebert and @nilsleiffischer on what that means for macOS users. I don't think we should require people use the latest version of macOS to build spectre.

**Comment by fmahebert** [comment]
I don't have a macOS spectre setup.

If someone else can test on 10.15 and/or 11, that's great.. if not I can work to set up a macOS env (I am still on 10.15).

**Comment by nilsvu** [comment]
LLVM clang-8 corresponds to AppleClang 11.0.0, which comes with macOS 10.15. I don't think anyone would be running an older version of macOS than that, though if they do they can always `brew install llvm`. Note that we haven't run CI on macOS for a while anyway, so we can't really guarantee anything but the versions that @kidder and I are running work.

**Comment by nilsdeppe** [comment]
Okay, sounds good. Can we add clang-11 and gcc-10 to CI?

**Comment by nilsvu** [comment]
Sure let's do that, though in another PR because it'll likely need some fixes. I have seen a lot of warnings pop up with AppleClang 12, mostly related to iterators that don't return by reference.

**Comment by nilsdeppe** [comment]
Sure, clang-11 and gcc-10 both build clean on my system, so this sounds like a macOS issue

**Comment by nilsdeppe** [comment]
I'm happy with dropping 6 & 7 and adding clang-11 and gcc-10 to CI in a separate PR. @wthrowe @kidder could you please share your thoughts? Thanks!

**Comment by wthrowe** [comment]
Sounds fine.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**